// @revolist/revogrid/dist/collection/components/header/headerService.js
import interact from 'interactjs';
import { DATA_COL, MIN_COL_SIZE } from '../../utils/consts';
export default class HeaderService {
    constructor(target, columns, config) {
        this.target = target;
        this.config = config;
        this.source = [];
        this.columns = columns;
        this.resizeChange(config.canResize);
    }
    get columns() {
        return this.source;
    }
    set columns(source) {
        this.source = source;
    }
    resizeChange(canResize = false) {
        this.destroy();
        if (canResize) {
            interact(this.target).resizable({
                edges: { bottom: false, right: true },
                onend: event => {
                    const index = parseInt(event.target.getAttribute(DATA_COL), 10);
                    const col = this.columns[index];
                    let width = event.rect.width;
                    const minSize = col.minSize || MIN_COL_SIZE;
                    if (width < minSize) {
                        width = minSize;
                    }
                    this.config.resize({ [index]: width });
                }
            });
        }
    }
    destroy() {
        interact(this.target).unset();
    }
}
